paasta_tools/marathon_tools.py

```diff
@@ -536,9 +547,19 @@ def get_host_port(self):
         '''Map this port on the host to your container's port 8888. Default is 0, which means Marathon picks a port.'''
         return self.config_dict.get('host_port', 0)
 
+    def get_marathon_shard(self):
+        """Returns the configued shard of Marathon to use.
+        Defaults to 0 currently as a fail-safe"""
+        return self.config_dict.get('marathon_shard', 0)
+
+    def get_previous_marathon_shards(self):
```

Comment:
I don't think this coupling is the right one - the shard isn't a part of the app - rather there should be another datastructure that encodes the pairing of a shard an an app owned by whatever is in charge of tying those things together,

Reply:
But it sure is convenient if we put the desired shard and the function that calculates the auto shard in here, as it can call self.get_marathon_shard() and the function for calculating the shard can inspect that and default to calculating based on self.service and self.instance. 

I'm not saying they could be separate functions not in this class, but we would have to inject service, instance into them in order to calculate the shard.

In other words, method 1:
```
shard = instanceconfig.get_marathon_shard(available_shards=shards)
app = marathon_tools.get_marathon_app(client=shards[shard], id=instanceconfig.generate_app_id()
```

Method 2:
```
prefered_shard = instanceconfig.get_preferred_marathon_shard()
if preferred_shard == None:
    prefrerred_shard = MarathonShardsThingy.calculate_shard_for_service(instanceconfig.service, instanceconfig.instance, available_shards=shards)
app = marathon_tools.get_marathon_app(client=shards[preferred_shard], id=instanceconfig.generate_app_id()
```

It isn't "that" much different. I would argue that the shard *is* kinda part of the app, because it is as much of a function of its name and config as it is a function of the shard topology.

Reply:
You're right that they dont appear that different, but I think that the presence of ``MarathonShardsThingy`` is a good one - we should be leaning towards the instance_config *just* being an expression of the data - not an expression of the data + a bunch of logic for calculating that data. It may seem small, but it significantly changes the overhead of 'what' instanceconfig is.

Saying that, the pattern is kind of established for instanceconfig doing this anyway (calculating the right value for its internal state, rather than something else being responsible for that,). This is just kind of kerbing that pattern where we can.